tests/test_orquestra_device.py

```diff
@@ -84,6 +100,112 @@ def test_empty_apply(self):
 
         assert dev.apply([]) is None
 
+    @pytest.mark.parametrize("keep", [True, False])
+    def test_keep_workflow_file(self, keep, tmpdir, monkeypatch):
+        """Test the option for keeping/deleting the workflow file."""
+
+        file_name = "test_workflow.yaml"
+        dev = qml.device("orquestra.forest", wires=3, keep_files=keep)
+        mock_res_dict = {"First": {"expval": {"list": [123456789]}}}
+        test_uuid = "1234"
+
+        assert not os.path.exists(tmpdir.join(f"expval-{test_uuid}.yaml"))
+        assert not dev.filenames
+        with monkeypatch.context() as m:
+            m.setattr(pennylane_orquestra.cli_actions, "user_data_dir", lambda *args: tmpdir)
+
+            # Disable submitting to Orquestra by mocking Popen
+            m.setattr(subprocess, "Popen", lambda *args, **kwargs: MockPopen())
+            m.setattr(
+                pennylane_orquestra.orquestra_device,
+                "loop_until_finished",
+                lambda *args, **kwargs: mock_res_dict,
+            )
+
+            # Disable random uuid generation
+            m.setattr(uuid, "uuid4", lambda *args: test_uuid)
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.PauliX(0)
+                return qml.expval(qml.PauliZ(0))
+
+            assert circuit() == 123456789
+            file_kept = os.path.exists(tmpdir.join(f"expval-{test_uuid}.yaml"))
+            assert file_kept if keep else not file_kept
+            assert dev.filenames == ([f"expval-{test_uuid}.yaml"] if keep else [])
+            assert dev.latest_id == "SomeWorkflowID"
+
+    @pytest.mark.parametrize("timeout", [1, 2.5])
+    def test_timeout(self, timeout, tmpdir, monkeypatch):
+        """Test the option for keeping/deleting the workflow file."""
+
+        file_name = "test_workflow.yaml"
+        dev = qml.device("orquestra.forest", wires=3, timeout=timeout)
+        mock_res_dict = {"First": {"expval": {"list": [123456789]}}}
+
+        test_uuid = "1234"
+        assert dev._timeout == timeout
+        assert not os.path.exists(tmpdir.join(f"expval-{test_uuid}.yaml"))
+        with monkeypatch.context() as m:
+            m.setattr(pennylane_orquestra.cli_actions, "user_data_dir", lambda *args: tmpdir)
+            m.setattr(pennylane_orquestra.cli_actions, "workflow_results", lambda *args: "Test res")
+
+            # Disable submitting to Orquestra by mocking Popen
+            m.setattr(subprocess, "Popen", lambda *args, **kwargs: MockPopen())
+
+            # Disable random uuid generation
+            m.setattr(uuid, "uuid4", lambda *args: test_uuid)
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.PauliX(0)
+                return qml.expval(qml.PauliZ(0))
+
+            start = time.time()
+            with pytest.raises(TimeoutError, match="The workflow results for workflow"):
+                circuit()
+            end = time.time()
+            assert end - start >= timeout
```

Comment:
Is this test simply reliant on that the calling of `circuit()` will take longer than `timeout` (in this case 1 and 2.5 seconds)? What if it's run on a very fast computer? Might it fail then? :thinking:

Reply:
The test is created so that [this while loop](https://github.com/PennyLaneAI/pennylane-orquestra/blob/a654fb207e1448f077f57b74dd44eac84bcb6e16/pennylane_orquestra/cli_actions.py#L182) is an infinite loop (due to an `IndexError` being raised everytime for `"Test res"` when a location is attempted to be parsed). Therefore raising the `TimeoutError` is interrupting the infinite loop.